src/hotspot/cpu/aarch64/c2_MacroAssembler_aarch64.cpp

```diff
@@ -2067,18 +2070,60 @@ void C2_MacroAssembler::neon_reduce_mul_integral(Register dst, BasicType bt,
         assert(false, "unsupported");
         ShouldNotReachHere();
     }
-  BLOCK_COMMENT("} neon_reduce_mul_integral");
+  BLOCK_COMMENT("} reduce_mul_integral_le128b");
+}
+
+// Vector reduction multiply for integral type with SVE instructions. Multiplies halves of the
+// source vector to get to a 128b vector that fits into a SIMD&FP register. After that point ASIMD
+// instructions are used. Note: temporary registers vtmp1 and vtmp2 are not used in some cases.
+// Clobbers: rscratch1
+void C2_MacroAssembler::reduce_mul_integral_gt128b(Register dst, BasicType bt, Register isrc,
+                                                   FloatRegister vsrc,
+                                                   unsigned vector_length_in_bytes,
+                                                   FloatRegister vtmp1, FloatRegister vtmp2,
+                                                   PRegister pgtmp) {
+  assert(vector_length_in_bytes > FloatRegister::neon_vl, "ASIMD impl should be used instead");
+  assert(vector_length_in_bytes <= FloatRegister::sve_vl_max, "unsupported vector length");
+  assert(is_power_of_2(vector_length_in_bytes), "unsupported vector length");
+
+  BLOCK_COMMENT("reduce_mul_integral_gt128b {");
+  unsigned vector_length = vector_length_in_bytes / type2aelembytes(bt);
+
+  auto do_recursive_folding_iteration =
+      [&](FloatRegister vdst, FloatRegister vsrc, FloatRegister vtmp) {
+        assert(vdst == vtmp || vdst == vsrc, "unsupported combination of registers");
+        sve_gen_mask_imm(pgtmp, bt, vector_length / 2);
+        // Shuffle the upper half elements of the register to the right.
+        sve_movprfx(vtmp1, vsrc);
+        sve_ext(vtmp1, vsrc, vector_length_in_bytes / 2);
+        if (vdst == vtmp) {
+          sve_mul(vdst, elemType_to_regVariant(bt), pgtmp, vsrc);
+        } else if (vdst == vsrc) {
+          sve_mul(vdst, elemType_to_regVariant(bt), pgtmp, vtmp);
+        } else {
+          ShouldNotReachHere();
+        }
+        vector_length_in_bytes = vector_length_in_bytes / 2;
+        vector_length = vector_length / 2;
+      };
+
+  do_recursive_folding_iteration(vtmp1, vsrc, vtmp1);
+  while (vector_length_in_bytes > FloatRegister::neon_vl) {
+    do_recursive_folding_iteration(vtmp1, vtmp1, vtmp2);
+  }
```

Comment:
Looks a little complex. Could we just simplify with following change? BTW, `sve_movprfx` inside of the loop can be saved. Please correct me if any mis-understanding!

```suggestion

  sve_movprfx(vtmp2, vsrc);
  while (vector_length_in_bytes > FloatRegister::neon_vl) {
    unsigned vector_length = vector_length_in_bytes / type2aelembytes(bt);
    sve_gen_mask_imm(pgtmp, bt, vector_length / 2);
    // Shuffle the upper half elements of the register to the right.
    sve_ext(vtmp1, vtmp2, vector_length_in_bytes / 2);
    sve_mul(vtmp2, elemType_to_regVariant(bt), pgtmp, vtmp1);
    vector_length_in_bytes = vector_length_in_bytes / 2;
  }
```

Reply:
> `    sve_ext(vtmp1, vtmp2, vector_length_in_bytes / 2);`

This doesn't look right. `vtmp1` is both an src and dest register operand for `EXT` but it contains an undefined value at the first iteration of the loop.

I agree that the implementation of the loop is confusing. I'll rework and (hopefully) simplify it, thank you for pointing this out.